Add support for NVTX functions.

ezyang opened the pull request:
This commit adds support for the simple, ASCII message only
NVTX functions.  There are a few more but these are the ones
I'm mostly interested in.

I tested by running this script under nvprof:

    import torch.cuda.nvtx
    torch.cuda.nvtx.range_push("foo")
    torch.cuda.nvtx.mark("bar")
    torch.cuda.nvtx.range_pop()

And verifying that the events showed up.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Add support for NVTX functions.

This commit adds support for the simple, ASCII message only
NVTX functions.  There are a few more but these are the ones
I'm mostly interested in.

I tested by running this script under nvprof:

    import torch.cuda.nvtx
    torch.cuda.nvtx.range_push("foo")
    torch.cuda.nvtx.mark("bar")
    torch.cuda.nvtx.range_pop()

And verifying that the events showed up.

NB: An earlier version of this patch lazily loaded libNvExtTools.so.
At the prompting of @colesbury, we decided against this, because
dynamically loading the library means we can get into bad
situations where we load the wrong version of the library.  If we
do it at build time, this situation won't occur.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

apaszke commented on setup.py:
```diff
@@ -325,7 +325,7 @@ def run(self):
     extra_compile_args += ['-DWITH_CUDA']
     extra_compile_args += ['-DCUDA_LIB_PATH=' + cuda_lib_path]
-    main_libraries += ['cudart']
+    main_libraries += ['cudart', 'nvToolsExt']
```
I don't think we should always be linking against nvtx. It'd be better to load it at runtime using ctypes

colesbury replied:
I think this is the right way to do it to ensure that you load the correct nvToolsExt library

apaszke reviewed (changes requested):
I'd be nice to add support for colors too. Should be easy to do

colesbury commented on torch/cuda/nvtx.py:
```diff
@@ -0,0 +1,48 @@
+    if lib is None:
+        # nvToolsExt.so is lazily loaded, so if you want to use these tools
+        # the library needs to be in your LD_LIBRARY_PATH
+        lib = ctypes.cdll.LoadLibrary("libnvToolsExt.so")
```
I don't think this is a good idea. The first solution was better. We distribute CUDA libraries with PyTorch. You don't want to load libnvToolsExt.so from a different CUDA installation.

apaszke replied:
Do we distribute all of them? I thought these are just the ones we use? If not, we should remove all unused libs, as it will make the binaries smaller.

colesbury replied:
No, but we should distribute libnvToolsExt.so. It's only 40 KiB. Otherwise you end up in the mess we had for cuDNN and libcudart

apaszke replied:
Eh ok

colesbury commented on torch/cuda/nvtx.py:
```diff
@@ -0,0 +1,46 @@
+import os
+import ctypes
+import warnings
+import torch.cuda
```
Remove the unused imports. It would also be good to import this from `torch/cuda/__init__.py`. That way the `torch.cuda.nvtx.xxx` functions are immediately available when `torch` is imported.

colesbury reviewed (commented):
A simple test would be good so at least we can check that the library is loaded correctly.

colesbury commented on torch/cuda/nvtx.py:
```diff
@@ -0,0 +1,46 @@
+    """
+    if _libnvToolsExt() is None:
+        raise RuntimeError('Unable to load nvToolsExt library')
+    return lib.nvtxMarkA(ctypes.c_char_p(msg))
```
`nvtxMarkA` returns `void`. I think you may need to set the `restype` in `_libnvToolsExt`. (The default is int -- I'm not sure what happens if you call a void function as if it returns int).

ezyang committed:
Handle review comments on nvtx support.

- Remove redundant imports from torch.cuda.nvtx
- Import torch.cuda.nvtx from torch.cuda, so it becomes available when
  you import torch.cuda
- Add a small, trivial test to test_cuda.py

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

colesbury commented:
This looks good, just fix the lint and the Python 3 contbuild

ezyang committed:
Fix lint errors.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Handle Python 3 bytes/Unicode distinction.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
All good!

apaszke commented on torch/cuda/nvtx.py:
```diff
@@ -0,0 +1,48 @@
+    global lib
+    if lib is None:
+        lib = ctypes.cdll.LoadLibrary(None)
+        lib.nvtxMarkA.restype = None
```
Don't `nvtxRangePushA` and `nvtxRangePopA` return void as well?

ezyang replied:
Nope, they return integers describing how many levels of nesting the ranges have.
